src/hotspot/share/gc/shenandoah/shenandoahScanRemembered.cpp

```diff
@@ -87,29 +87,200 @@ void ShenandoahDirectCardMarkRememberedSet::merge_overreach(size_t first_cluster
 ShenandoahScanRememberedTask::ShenandoahScanRememberedTask(ShenandoahObjToScanQueueSet* queue_set,
                                                            ShenandoahObjToScanQueueSet* old_queue_set,
                                                            ShenandoahReferenceProcessor* rp,
-                                                           ShenandoahRegionIterator* regions) :
+                                                           ShenandoahRegionChunkIterator* work_list, bool is_concurrent) :
   WorkerTask("Scan Remembered Set"),
-  _queue_set(queue_set), _old_queue_set(old_queue_set), _rp(rp), _regions(regions) {}
+  _queue_set(queue_set), _old_queue_set(old_queue_set), _rp(rp), _work_list(work_list), _is_concurrent(is_concurrent) {}
 
 void ShenandoahScanRememberedTask::work(uint worker_id) {
-  // This sets up a thread local reference to the worker_id which is necessary
-  // the weak reference processor.
-  ShenandoahParallelWorkerSession worker_session(worker_id);
+  if (_is_concurrent) {
+    // This sets up a thread local reference to the worker_id which is needed by the weak reference processor.
+    ShenandoahConcurrentWorkerSession worker_session(worker_id);
+    ShenandoahSuspendibleThreadSetJoiner stsj(ShenandoahSuspendibleWorkers);
+    do_work(worker_id);
+  } else {
+    // This sets up a thread local reference to the worker_id which is needed by the weak reference processor.
+    ShenandoahParallelWorkerSession worker_session(worker_id);
+    do_work(worker_id);
+  }
+}
+
+void ShenandoahScanRememberedTask::do_work(uint worker_id) {
   ShenandoahWorkerTimingsTracker x(ShenandoahPhaseTimings::init_scan_rset, ShenandoahPhaseTimings::ScanClusters, worker_id);
 
   ShenandoahObjToScanQueue* q = _queue_set->queue(worker_id);
   ShenandoahObjToScanQueue* old = _old_queue_set == NULL ? NULL : _old_queue_set->queue(worker_id);
   ShenandoahMarkRefsClosure<YOUNG> cl(q, _rp, old);
-  RememberedScanner* scanner = ShenandoahHeap::heap()->card_scan();
+  ShenandoahHeap* heap = ShenandoahHeap::heap();
+  RememberedScanner* scanner = heap->card_scan();
 
   // set up thread local closure for shen ref processor
   _rp->set_mark_closure(worker_id, &cl);
-  ShenandoahHeapRegion* region = _regions->next();
-  while (region != NULL) {
-    log_debug(gc)("ShenandoahScanRememberedTask::work(%u), looking at region " SIZE_FORMAT, worker_id, region->index());
+  struct ShenandoahRegionChunk assignment;
+  bool has_work = _work_list->next(&assignment);
+  while (has_work) {
+#ifdef ENABLE_REMEMBERED_SET_CANCELLATION
+    // This check is currently disabled to avoid crashes that occur
+    // when we try to cancel remembered set scanning
+    if (heap->check_cancelled_gc_and_yield(_is_concurrent)) {
+      return;
+    }
+#endif
```

Comment:
The crashes are likely because you are pulling an assignment off the worklist, and if you see a cancellation, you are dropping it on the floor and returning.

I think the check should be at the end of the loop after the assignment pulled off the work list has been processed but before the next one is pulled.

Here's a diff that should work:

```
diff --git a/src/hotspot/share/gc/shenandoah/shenandoahScanRemembered.cpp b/src/hotspot/share/gc/shenandoah/shenandoahScanRemembered.cpp
index 2d7cbbb97dd..03f647f23b2 100644
--- a/src/hotspot/share/gc/shenandoah/shenandoahScanRemembered.cpp
+++ b/src/hotspot/share/gc/shenandoah/shenandoahScanRemembered.cpp
@@ -81,15 +81,7 @@ void ShenandoahScanRememberedTask::do_work(uint worker_id) {
   // set up thread local closure for shen ref processor
   _rp->set_mark_closure(worker_id, &cl);
   struct ShenandoahRegionChunk assignment;
-  bool has_work = _work_list->next(&assignment);
-  while (has_work) {
-#ifdef ENABLE_REMEMBERED_SET_CANCELLATION
-    // This check is currently disabled to avoid crashes that occur
-    // when we try to cancel remembered set scanning
-    if (heap->check_cancelled_gc_and_yield(_is_concurrent)) {
-      return;
-    }
-#endif
+  while (_work_list->next(&assignment)) {
     ShenandoahHeapRegion* region = assignment._r;
     log_debug(gc)("ShenandoahScanRememberedTask::do_work(%u), processing slice of region "
                   SIZE_FORMAT " at offset " SIZE_FORMAT ", size: " SIZE_FORMAT,
@@ -101,13 +93,15 @@ void ShenandoahScanRememberedTask::do_work(uint worker_id) {
       assert(clusters * cluster_size == assignment._chunk_size, "Chunk assignments must align on cluster boundaries");
       HeapWord* end_of_range = region->bottom() + assignment._chunk_offset + assignment._chunk_size;
 
-      // During concurrent mark, region->top() equals TAMS with respect to the current young-gen pass.  */
+      // During concurrent mark, region->top() equals TAMS with respect to the current young-gen pass.
       if (end_of_range > region->top()) {
         end_of_range = region->top();
       }
       scanner->process_region_slice(region, assignment._chunk_offset, clusters, end_of_range, &cl, false, _is_concurrent);
     }
-    has_work = _work_list->next(&assignment);
+    if (heap->check_cancelled_gc_and_yield(_is_concurrent)) {
+      return;
+    }
   }
 }
```

Reply:
Thanks very much for figuring out this problem.  I'd like to address this change in a different pull request to keep the history cleaner.

Reply:
👍

Reply:
I left a SPECjbb run overnight with the changes, but alas it still does crash (slowdebug mode) during a deoptimization, presumably because of a bad oop on the stack. I'll file a ticket to track the issue, but it does seem like there's some other issue besides the one pointed out above.